Framework/src/CheckRunner.cxx

```diff
@@ -319,6 +322,32 @@ void CheckRunner::send(QualityObjectsType& qualityObjects, framework::DataAlloca
   }
 }
 
+void CheckRunner::updateServiceDiscovery(const QualityObjectsType& qualityObjects)
+{
+  if (mServiceDiscovery == nullptr) {
+    return;
+  }
+
+  // insert into the list of paths the QOs' paths.
+  // TODO it seems unoptimal to do it all the time. After a while, aren't we sure that we have run all the checks ?
```

Comment:
Don't we actually know all the QO names during initialization? Why doing it once is not enough?

Reply:
Right, I got mistaken with the new policy `OnEachSeparately`. Actually each Check has either 1 single output or as many outputs as inputs in case of `OnEachSeparately`, correct ? 

I'll rework this part.

Reply:
Hmm, in case of `OnEachSeparately` it is as many as there are MOs. But notice that if someone declares "all" MOs, then you don't know the amount of them. So in that case indeed we don't know the number of QOs during the initializaiton.
So maybe it is actually worth to leave it like that. Another approach for `OnEachSeparately` would be to declare only the Check, not the QOs themselves, then we could do it once, during init.

Reply:
Ok, just remembered what was the issue. If we use `OnEachSeparately` and the config file says 
```
{
          "type": "Task",
          "name": "TaskName",
          "MOs": "all"
        }
```
then we don't know what are the MOs and thus cannot build the corresponding paths during initialization. 

Am I missing something ?

Reply:
Yes, exactly.
So you were right to have the update function inside `run()`

Reply:
Added some info to remember why it is like that and how to improve it if we want.